daemon/c_seccomp.c

```diff
@@ -0,0 +1,534 @@
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GPL 2 license for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, see <http://www.gnu.org/licenses/>
+ *
+ * The full GNU General Public License is included in this distribution in
+ * the file called "COPYING".
+ *
+ * Contact Information:
+ * Fraunhofer AISEC <gyroidos@aisec.fraunhofer.de>
+ */
+
+/**
+ * @file c_seccomp.c
+ *
+ * This module handles system call hooking and emulation through the seccomp notify kernel functionality.
+ * It applies a seccomp filter matching relevant system calls before the container init process is executed.
+ * Access to global resources on behalf of a container through this module is controlled by flags
+ * stored in the 'enabled_features' variable. If access to a particular resource is allowed,
+ * this module emulates the syscalls a container issues to access the resource.
+ */
+
+#define _GNU_SOURCE
+
+#define MOD_NAME "c_seccomp"
+
+#include "cmld.h"
+#include "compartment.h"
+#include "audit.h"
+
+#include "common/macro.h"
+#include "common/mem.h"
+#include "common/fd.h"
+#include "common/event.h"
+#include "common/audit.h"
+
+#include <unistd.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <sys/ioctl.h>
+#include <sys/syscall.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <sys/timex.h>
+
+#include <linux/filter.h>
+#include <linux/seccomp.h>
+#include <linux/audit.h>
+#include <linux/netlink.h>
+
+//#undef LOGF_LOG_MIN_PRIO
+//#define LOGF_LOG_MIN_PRIO LOGF_PRIO_TRACE
+
+// TODO multiarch support?
+#define X32_SYSCALL_BIT 0x40000000
+
+//#define __NR_pidfd_open		434
+//#define __NR_pidfd_getfd 	438
+
+typedef struct c_seccomp {
+	compartment_t *compartment;
+	struct seccomp_notif_sizes *notif_sizes;
+	event_io_t *event;
+	unsigned int enabled_features;
+} c_seccomp_t;
+
+static bool
+seccomp_is_pidfd_supported()
+{
+	return cmld_kernel_version_check("5.10.0");
+}
+
+int
+seccomp(unsigned int op, unsigned int flags, void *args)
+{
+	errno = 0;
+	return syscall(__NR_seccomp, op, flags, args);
+}
+
+int
+seccomp_ioctl(int fd, unsigned long request, void *notify_req)
+{
+	errno = 0;
+	return syscall(__NR_ioctl, fd, request, notify_req);
+}
+
+#define NCALLS 2
+
+// slightly modified sample code from the seccomp manpage
+// https://man7.org/linux/man-pages/man2/seccomp.2.html
+// an the Linux secomp sample code at samples/seccomp
+int
+install_filter(unsigned int t_arch, UNUSED int f_errno)
+{
+	unsigned int upper_nr_limit = 0xffffffff;
+
+	/* Assume that AUDIT_ARCH_X86_64 means the normal x86-64 ABI
+     * (in the x32 ABI, all system calls have bit 30 set in the
+     *   'nr' field, meaning the numbers are >= X32_SYSCALL_BIT)
+     *
+     */
+	if (t_arch == AUDIT_ARCH_X86_64)
+		upper_nr_limit = X32_SYSCALL_BIT - 1;
+
+	struct sock_filter filter[] = {
+		/* [0] Load architecture from 'seccomp_data' buffer into
+                accumulator */
```

Comment:
format of comments should be applied here too

Reply:
done